apps/builder/app/shared/sync-client.ts

```diff
@@ -0,0 +1,127 @@
+import { nanoid } from "nanoid";
+import { createNanoEvents, type Emitter } from "nanoevents";
+import type { Change, Store } from "immerhin";
+
+type Transaction = {
+  id: string;
+  changes: Change[];
+};
+
+type SyncMessage =
+  | { type: "connect"; clientId: string }
+  | { type: "state"; clientId: string; state: Map<string, unknown> }
+  | { type: "commit"; clientId: string; transaction: Transaction }
+  | { type: "revert"; clientId: string; transactionId: string };
+
+export type SyncEmitter = Emitter<{
+  message: (message: SyncMessage) => void;
+}>;
+
+type SyncClientOptions = {
+  role: "leader" | "follower";
+  store: Store;
+  emitter?: SyncEmitter;
+};
+
+export class SyncClient {
+  clientId = nanoid();
```

Comment:
do you need a nanoid generator here or could be just a simple counter or a Math.random call without dependencies?

Reply:
I need unique id to communicate with other tabs or users.

Reply:
Currently, clientId is used to district own from received messages.